### PR TITLE
Clarify that match() & like() use UTF-8 for matching

### PR DESCRIPTION
**Pull request by rschu1ze**
The previous sentence in the documentation of match() and like()
```
  "The regular expression works with the string as if it is a set of bytes."
```
suggested matching is based on Latin-1/ASCII. Since we don't have separate methods `matchUTF8()` and `likeUTF8()` (like we have e.g. with `position()`/`positionUTF8()`), we should clarify matching is really based on UTF-8.

### Changelog category (leave one):
- Documentation (changelog entry is not required)

### CODE REVIEW

**Comment by rschu1ze** [comment]
@Mergifyio update

(just to re-trigger "Docs Check" which failed with

```
++ git branch --show-current
+ GIT_BRANCH=main
+ '[' '' ']'
+ git pull
From https://github.com/ClickHouse/clickhouse-docs
   a7cebef..b1cdd27  main       -> origin/main
Updating a7cebef..b1cdd27
error: The following untracked working tree files would be overwritten by merge:
	yarn.lock
Please move or remove them before you merge.
Aborting
```
)

**Comment by rschu1ze** [comment]
StyleCheck complains about trailing whitespaces in ReadFromRemote.cpp. Tried to fix it by myself but GitHub rejected my push because of missing permissions. Would you like to fix the whitespaces too? Thanks.

**Comment by rschu1ze** [comment]
Meh, that didn't work.

**Comment by alexey-milovidov** [comment]
Yes, you can say it's UTF-8, but there are many details:

```
ip-172-31-14-26.eu-central-1.compute.internal :) SELECT 'ё' AS yo1, normalizeUTF8NFKD('ё') AS yo2, upperUTF8(yo1) AS Yo1, upperUTF8(yo2) AS Yo2, yo1 LIKE '%' || yo2 || '%', match(yo1, yo2), match(yo1, '(?i)' || Yo1), yo2 LIKE '%\xB5\xCC%'

SELECT
    'ё' AS yo1,
    normalizeUTF8NFKD('ё') AS yo2,
    upperUTF8(yo1) AS Yo1,
    upperUTF8(yo2) AS Yo2,
    yo1 LIKE concat('%', yo2, '%'),
    match(yo1, yo2),
    match(yo1, concat('(?i)', Yo1)),
    yo2 LIKE '%��%'

Query id: 12e0d78f-0167-4b7c-b453-0b242e76e74e

┌─yo1─┬─yo2─┬─Yo1─┬─Yo2─┬─like('ё', concat('%', normalizeUTF8NFKD('ё'), '%'))─┬─match('ё', normalizeUTF8NFKD('ё'))─┬─match('ё', concat('(?i)', upperUTF8('ё')))─┬─like(normalizeUTF8NFKD('ё'), '%��%')─┐
│ ё   │ ё   │ Ё   │ Ё   │                                                   0 │                                  0 │                                          1 │                                  1 │
└─────┴─────┴─────┴─────┴─────────────────────────────────────────────────────┴────────────────────────────────────┴────────────────────────────────────────────┴────────────────────────────────────┘
```

1. Case-insensitive match with general collation algorithm (but not locale-aware algorithm like Turkish)
2. Unicode composition/normalization.
3. Some functions may work on invalid UTF-8 sequences in some cases, which is desirable behavior.

File: docs/en/sql-reference/functions/string-search-functions.md [hunk 1]
```diff
@@ -350,11 +350,12 @@ In all `multiSearch*` functions the number of needles should be less than 2<sup>
 
 ## match(haystack, pattern) {#matchhaystack-pattern}
 
-Checks whether the string matches the `pattern` regular expression. A `re2` regular expression. The [syntax](https://github.com/google/re2/wiki/Syntax) of the `re2` regular expressions is more limited than the syntax of the Perl regular expressions.
+Checks whether the string matches the regular expression `pattern` in `re2` syntax. `Re2` has a more limited [syntax](https://github.com/google/re2/wiki/Syntax) than Perl regular expressions.
 
 Returns 0 if it does not match, or 1 if it matches.
 
-The regular expression works with the string as if it is a set of bytes. The regular expression can’t contain null bytes.
+Matching is based on UTF-8, e.g. `.` matches the two-codepoint symbol `¥`. The regular expression must not contain null bytes.
```
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/string-search-functions.md:357]: No, ¥ is one Unicode code point (U+00A5) that is encoded by two bytes in UTF-8.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: docs/en/sql-reference/functions/string-search-functions.md (hunk 1 above)
Comment (alexey-milovidov) [reply docs/en/sql-reference/functions/string-search-functions.md:357]: grapheme cluster > code point > byte

**Comment by rschu1ze** [comment]
Agree, that the devil is in the details. In fact, the complexity of Unicode is so high that a detailed specification of the exact behavior of each `*UTF8()` function variant in ClickHouse is not really realistic. We would need to mention besides the issues you mentioned the internally supported Unicode standard (the latest is 14.0), treatment of BOMs in UTF8, handling of private use code points, granularity of graphemes (e.g. flags) vs. code-point and probably a lot more.

But since ClickHouse comes with NF(K)(C/D) normalization and the general vs. locale-specific collation (Turkish i/İ vs i/I) is mentioned also elsewhere in the docs, let me add another sentence about these.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**